# DTSC 580: Data Manipulation
## Assignment: LEGO Project
### Name: 

![LegoSchema](Lego_Schema.PNG)

## Overview

In this assignment, you will load the various tables from the [Rebrickable](https://rebrickable.com/) website and perform analysis on the data.  Rebrickable is a fantastic website that is used to show you which LEGO sets you can build from the sets and parts you already own.  They will even show you which parts you are missing to complete the new set and suggest sets that you could purchase to get those parts.  

## Data

The data has been downloaded directly from their website and includes files for the tables shown in the schema above.  Your first task will be to go through the data and review the schema to understand how the tables all fit together.  There is no data dictionary that I am aware of, but you should be able to understand this data by reviewing the files and schema and visiting their [Download](https://rebrickable.com/downloads/) page and [API](https://rebrickable.com/api/) documentation.  **Note that you must use the files downloaded from Brightspace for this assignment and not download new files from the Rebrickable website.**

## Assignment

In this assignment, you will create a number of functions to complete certain tasks.  This is not needed if you were working on this project on your own, but since we want to see if you can select the data via Pandas/Numpy, we will use those functions in the automatic grading in CodeGrade.  Without this, students could try to look up the answers in the `csv` files instead of wrangling the data with Python/Pandas, which would defeat the purpose of this class.

In addition, we would strongly suggest that you merge data sets as appropriate for the given task.  For example, you might merge the `sets` and `themes` DataFrames to answer one question, and then merge `sets`, `inventories`, `inventory_minifigs`, and `minifigs` to answer another.  This will make your code much easier to write and will be key in working with data across multiple tables.  In fact, I was able to create each function with only one line of code because of the merging of certain data sets.  Also note that you may not use all the tables in this assignment.  This is also an important skill for a Data Scientist to know when to use data and what data is unnecessary.

When complete, save your notebook as `lego.ipynb` and submit to CodeGrade for automatic grading.

## Note

<u>Show Work</u>

Remember that you must show your work.  Students submissions are spot checked manually throughout the term to verify that they are not hard coding the answer from looking only in the file or in CodeGrade's expected output.  If this is seen, the student's answer will be manually marked wrong and their grade will be changed to reflect this. 

For example, if the answer to Q1, the mean of a specific column, is 22:
```
# correct way
Q1 = df['column_name'].mean()

# incorrect way
Q1 = 22 
```

Let's start out by importing some standard imports.

In [1]:
# common imports
import pandas as pd
import numpy as np

# Do not change this option; This allows the CodeGrade auto grading to function correctly
pd.set_option('display.max_columns', None)

## Import Data

Load the various Lego datasets below calling them the same as the name of the table as seen in the schema.  Note that you may or may not have to use all the data sets.  One of the things that all Data Scientists and Analysts have to decide when working on a project is what data is useful for a task and what data is not.  Remember to look at the schema above to see how all the data sets are connected.

The autograder in CodeGrade will have the `csv` files in the same location where the notebook is run.  Because of this, please make sure your local files are in the same location as your notebook.

In [2]:
# import Lego data sets here
# make sure that files are in the same folder as your notebook
inventories = pd.read_csv('inventories.csv')
inventory_sets = pd.read_csv('inventory_sets.csv')
sets = pd.read_csv('sets.csv')
themes = pd.read_csv('themes.csv')
inventory_minifigs = pd.read_csv('inventory_minifigs.csv')
minifigs = pd.read_csv('minifigs.csv')
inventory_parts = pd.read_csv('inventory_parts.csv')
part_categories = pd.read_csv('part_categories.csv')
parts = pd.read_csv('parts.csv')
colors = pd.read_csv('colors.csv')
part_relationships = pd.read_csv('part_relationships.csv')
elements = pd.read_csv('elements.csv')

#for my own use

#inventories = pd.read_csv('C:/Users/Public/inventories.csv')
#inventory_sets = pd.read_csv('C:/Users/Public/inventory_sets.csv')
#sets = pd.read_csv('C:/Users/Public/sets.csv')
#themes = pd.read_csv('C:/Users/Public/themes.csv')
#inventory_minifigs = pd.read_csv('C:/Users/Public/inventory_minifigs.csv')
#minifigs = pd.read_csv('C:/Users/Public/minifigs.csv')
#inventory_parts = pd.read_csv('C:/Users/Public/inventory_parts.csv')
#part_categories = pd.read_csv('C:/Users/Public/part_categories.csv')
#parts = pd.read_csv('C:/Users/Public/parts.csv')
#colors = pd.read_csv('C:/Users/Public/colors.csv')
#part_relationships = pd.read_csv('C:/Users/Public/part_relationships.csv')
#elements = pd.read_csv('C:/Users/Public/elements.csv')

## Exploration

Take time to look at all of the data sets using `.head()`, `.info()`, `.describe()`, etc to familiarize yourself with the data sets.  As you look at them, notice how the various tables connect together.

In [75]:
inventories
inventories.head()
#inventories.info()
#inventories.describe()


id  version set_num
0   1        1  7922-1
1   3        1  3931-1
2   4        1  6942-1
3  15        1  5158-1
4  16        1   903-1

In [77]:
inventory_sets
inventory_sets.head() #connected to inventories by set_num
#inventory_sets.info()
#inventory_sets.describe()

inventory_id  set_num  quantity
0            35  75911-1         1
1            35  75912-1         1
2            39  75048-1         1
3            39  75053-1         1
4            50   4515-1         1

In [78]:
sets
sets #connected to inventory_sets by set_num
#sets.info()
#sets.describe()

set_num                             name  year  theme_id  num_parts
0            001-1                            Gears  1965         1         43
1           0011-2                Town Mini-Figures  1979        67         12
2           0011-3       Castle 2 for 1 Bonus Offer  1987       199          0
3           0012-1               Space Mini-Figures  1979       143         12
4           0013-1               Space Mini-Figures  1979       143         12
...            ...                              ...   ...       ...        ...
18571      XWING-1              Mini X-Wing Fighter  2019       158         60
18572      XWING-2                X-Wing Trench Run  2019       158         52
18573  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158        413
18574   YTERRIER-1                Yorkshire Terrier  2018       598          0
18575     ZX8000-1             ZX 8000 LEGO Sneaker  2020       501          0

[18576 rows x 5 columns]

In [79]:
themes
themes #connected to inventory by id and connected to sets by name
#themes.info()
#themes.describe()

id                        name  parent_id
0      1                     Technic        NaN
1      3                 Competition        1.0
2      4              Expert Builder        1.0
3     16                  RoboRiders        1.0
4     17              Speed Slammers        1.0
..   ...                         ...        ...
439  715                      Marvel      535.0
440  716                     Modulex        NaN
441  717                 Speed Racer        NaN
442  718       Series 22 Minifigures      535.0
443  719  BrickLink Designer Program        NaN

[444 rows x 3 columns]

In [81]:
inventory_minifigs
inventory_minifigs.head() #connected to inventory_sets by inventory_id
#inventory_minifigs.info()
#inventory_minifigs.describe()

inventory_id     fig_num  quantity
0             3  fig-001549         1
1             4  fig-000764         1
2            19  fig-000555         1
3            25  fig-000574         1
4            26  fig-000842         1

In [82]:
inventory_parts
inventory_parts.head() #connected to inventory_minifigs and inventory_sets by inventory_id and quantity
#inventory_parts.info()
#inventory_parts.describe()

inventory_id        part_num  color_id  quantity is_spare
0             1        48379c01        72         1        f
1             1           48395         7         1        f
2             1  stickerupn0077      9999         1        f
3             1         upn0342         0         1        f
4             1         upn0350        25         1        f

In [83]:
part_categories
part_categories.head() # connected to themes and inventories by id, connected to themes and sets by name
#part_categories.info()
#part_categories.describe()

id                     name
0   1               Baseplates
1   3            Bricks Sloped
2   4  Duplo, Quatro and Primo
3   5           Bricks Special
4   6            Bricks Wedged

In [84]:
parts
parts.head() # conncted to themes, sets and part_categories by name, cntd inventory_parts by part_num, 

part_num                                        name  part_cat_id  \
0   003381                 Sticker Sheet for Set 663-1           58   
1   003383         Sticker Sheet for Sets 618-1, 628-2           58   
2   003402  Sticker Sheet for Sets 310-3, 311-1, 312-3           58   
3   003429                Sticker Sheet for Set 1550-1           58   
4   003432  Sticker Sheet for Sets 357-1, 355-1, 940-1           58   

  part_material  
0       Plastic  
1       Plastic  
2       Plastic  
3       Plastic  
4       Plastic

In [85]:
colors
colors.head() #connected to part_categories,themes,and inventories by 
#'id',connceted by parts, by parts,part_categories,themes,sets, 'name'
#colors.info()
#colors.describe()

id            name     rgb is_trans
0  -1       [Unknown]  0033B2        f
1   0           Black  05131D        f
2   1            Blue  0055BF        f
3   2           Green  237841        f
4   3  Dark Turquoise  008F9B        f

In [86]:
part_relationships
part_relationships.head()
#part_relationships.info()
#part_relationships.describe()

rel_type child_part_num parent_part_num
0        P    87079pr9974           87079
1        P     3960pr9971            3960
2        R    98653pr0003     98086pr0003
3        R    98653pr0003    98088pat0003
4        R    98653pr0003    98089pat0003

In [87]:
elements
elements.head() #conncted to parts,inventory_parts by part_num
#elements.info()
#elements.describe()

element_id  part_num  color_id
0    6300211  67906c01        14
1    4566309      2564         0
2    4275423     53657      1004
3    6194308     92926        71
4    6229123     26561         4

In [188]:
minifigs

fig_num                                   name  num_parts
0      fig-000001                     Toy Store Employee          4
1      fig-000002                           Customer Kid          4
2      fig-000003                  Assassin Droid, White          8
3      fig-000004                           Basic Figure          4
4      fig-000005        Captain America with Short Legs          3
...           ...                                    ...        ...
11822  fig-012221                                Bob-omb         11
11823  fig-012222                   Boomer (Banzai) Bill         10
11824  fig-012223                            Bullet Bill         10
11825  fig-012224                           Buzzy Beetle          7
11826  fig-012225  Unikitty - Cheerykitty (Cheery Kitty)         14

[11827 rows x 3 columns]

## Assignment

Answer all the questions below using the available data.  There are multiple ways to get some of the answers.  I would suggest that you attempt to merge some data sets as needed to assist with getting answers.  This will make your tasks a lot easier and your code will be cleaner.

Also, you will be creating functions to get the answers to some of these questions.  This will make it much easier for you to get answers to similar questions in the future and CodeGrade will be checking your functions using different inputs.  

**Q1:** Create a function called `data_count` that takes as input the data set (for example: `sets`) and returns the total number of rows in the data (as an integer).  Nothing else should be returned.

In [170]:
def data_count(x):
    return len(x.index)

**Code Check:**  If you call your `data_count` function using the `sets` DataFrame, you should see that there are 18,576 sets represented in the data.

In [171]:
data_count(sets)

18576

**Q2:** What is the average number of parts for a Lego set (rounded to the nearest integer)?  Save this as `Q2`.

In [138]:
#sets -> num_parts
Q2 = round(sets['num_parts'].mean())
Q2

161

**Q3:** What is the median number of parts per set (output as an integer)?  Save this as `Q3`.

In [139]:
# sets -> num_parts
Q3 = round(sets['num_parts'].median())
Q3

37

You should see that the median number of parts is much lower than the average number of parts per set.  This tells us that there are some Lego sets with a very high number of parts that brings the average up.

**Q4:** What is the largest number of parts that a set has? Save this as `Q4`.

In [140]:
Q4 = sets['num_parts'].max()
Q4

11695

**Q5:** Given the set with the largest number of parts calculated above, select its row from the `sets` DataFrame.  Save this as `Q5`.  Do not reset the index.

In [141]:
Q5 = sets[sets['num_parts'] == 11695]
Q5

set_num       name  year  theme_id  num_parts
3867  31203-1  World Map  2021       709      11695

**Q6:** Create a function called `select_set_row` where you pass it the `set_num` (as a string) and the function returns the respective row of the DataFrame from `sets`. Do not reset the index.

In [142]:
sets.head()

set_num                        name  year  theme_id  num_parts
0   001-1                       Gears  1965         1         43
1  0011-2           Town Mini-Figures  1979        67         12
2  0011-3  Castle 2 for 1 Bonus Offer  1987       199          0
3  0012-1          Space Mini-Figures  1979       143         12
4  0013-1          Space Mini-Figures  1979       143         12

In [143]:
#sets ->set_num , refer to Q5 for similiar set up
def select_set_row(set_num):
    return sets[sets['set_num'] == set_num]

In [144]:
# select_set_row('0011-3') check the function

**Q7:** Create a function called `select_set_numparts` that takes as input a specific number of parts (as an integer) and selects all the rows in the `sets` DataFrame that contain those number of parts.  Return the DataFrame sorted by `year` and then by `set_num`.  Do not reset the index.

In [145]:
#num_parts-> sets
def select_set_numparts(num_parts):
    df = sets[sets['num_parts'] == num_parts]
    return df.sort_values(by=['year','set_num'])

**Code Check:** Call the `select_set_numparts()` function to select all sets with `0` number of parts.  There should be 3,146 rows returned.

In [146]:
select_set_numparts(0)

set_num                                               name  \
11061          700.1-2                     Extra-Large Gift Set (Mursten)   
17563         b55dk-01               Lego Mursten - System i Leg Byggebog   
17564         b56de-01                                    System im Spiel   
11072          700.2-2                                       Gift Package   
11078         700.3A-2                                       Gift Package   
...                ...                                                ...   
15619         854195-1                                    Ernie Key Chain   
15620         854197-1                               Demogorgon Key Chain   
17170  9780744054583-1  DC Comics Super Heroes: Character Encyclopedia...   
17193  9780794447557-1                               City: Color The City   
17200  9780794448172-1           DC Comics Super Heroes: Ready For Action   

       year  theme_id  num_parts  
11061  1953       366          0  
17563  1955       497          0  
17564  1956       497          0  
11072  1957       366          0  
11078  1957       372          0  
...     ...       ...        ...  
15619  2022       503          0  
15620  2022       503          0  
17170  2022       497          0  
17193  2022       497          0  
17200  2022       497          0  

[3146 rows x 5 columns]

**Q8:** 
- The oldest set listed was made in what year?  Save the year as `Q8A` (as an integer).  
- The newest set listed was made in what year? Save the year as `Q8B` (as an integer).

In [147]:
#min
Q8A = sets['year'].min()
Q8A

1949

In [ ]:
#max
Q8B =sets['year'].max()
Q8B

**Q9:** 
- Create a function called `select_set_year` that takes as input a year (as an integer) and returns the rows from the `sets` DataFrame that matches that year.  Sort the DataFrame by `set_num`.  Do not reset the index.

In [ ]:
def select_set_year(year):
    df = sets[sets['year'] == year]
    return df.sort_values(by=['set_num'])

In [ ]:
select_set_year(1960)

**Q10:** 
- Create a function called `theme_by_year` that takes as input a year (as an integer) and shows the theme ids and theme names (listed in order by theme id) that were in sets that year.  
- The column names must be `id` and `name_themes` (to differentiate between the name of a theme and the name of a set) in that order.  
- The index should be reset and go from `0` to `n-1`.  
- Each theme should only be listed once even if it appeared in more than one set from that year -- duplicate themes should be based on theme id and not name since there are some themes with the same name but with a different id.
- Hint:  It will help if you were to think about merging appropriate DataFrames to help you get this answer.

In [ ]:
sets

In [ ]:
themes

In [148]:
#year
def theme_by_year(year):
    #rename the columns in the themes column to be able to merge with the sets column
    data_rename = themes.rename(columns={"id": "theme_id", "name":"name_themes"})
    years = sets[sets['year'] == year]
    data_merge = pd.merge(data_rename, years, on='theme_id')
    #remove duplicates 
    unique_data = data_merge[['theme_id', 'name_themes']].drop_duplicates()
    #change back the theme_id to id
    unique_name = unique_data.rename(columns={'theme_id':'id'})
    data_reset = unique_name.reset_index(drop=True)
    return data_reset


**Code Check:** Call the `theme_by_year()` function on the year `1960`.  Your output should look as follows:

| | id | name_themes|
|-|---|---|
|0|371|Supplemental|
|1|497|Books|
|2|513|Classic|

**Optional Practice:** Call the `theme_by_year()` function using your birth year to see the Lego themes that were popular when you were born.

In [149]:
theme_by_year(1960)

id   name_themes
0  371  Supplemental
1  497         Books
2  513       Classic

**Q11:** 
- Create a function called `theme_by_name` that takes as input a theme name and outputs all the sets (in `set_num` order) that had that theme.  
- The DataFrame should include the following columns in this order and with these exact names: `set_num`,`name_sets`,`year`,`num_parts`, `theme_id`, and `name_themes`.  
- The index should go from `0` to `n-1`.

In [ ]:
sets

In [150]:
def theme_by_name(theme_name):
    #rename the theme id and name columns to appropiate
    rename_themes = themes.rename(columns={'id':'theme_id', 'name':'name_themes'})
    #rename set name column to name_set
    rename_sets = sets.rename(columns={'name':'name_sets'})
    #merge columns
    data_merged = pd.merge(rename_sets,rename_themes, on='theme_id')
    #filter by name_themes
    please = data_merged[data_merged['name_themes'] == theme_name]
    #choose column order and reset index
    please_work = please[['set_num', 'name_sets', 'year', 'num_parts', 'theme_id', 'name_themes']].sort_values(by='set_num')
    answer = please_work.reset_index(drop=True)
    return answer

**Code Check:** Call the `theme_by_name()` function using the theme name `Gear`.  You should return 1,904 sets that match that theme name.

In [151]:
theme_by_name('Gear')

set_num                                          name_sets  \
0              11138-1                            Bionicle Barraki Wallet   
1       119645111436-1                Classic Space Astronaut Plush (Red)   
2       119645111504-1               Classic Space Astronaut Plush (Blue)   
3       119645111740-1              Classic Space Astronaut Plush (White)   
4       119645111818-1               Classic Space Astronaut Plush (Pink)   
...                ...                                                ...   
1899           WATCH-3                   Watch System Voyager (UFO Space)   
1900         WIIGAME-1                              Bionicle Heroes - Wii   
1901         WIIGAME-2  Indiana Jones: The Original Adventures Video G...   
1902  WOODENELEPHANT-1  Original Wooden Elephant Pull-Along Toy and Print   
1903          ZX8000-1                               ZX 8000 LEGO Sneaker   

      year  num_parts  theme_id name_themes  
0     2008          0       501        Gear  
1     2021          0       501        Gear  
2     2021          0       501        Gear  
3     2021          0       501        Gear  
4     2021          0       501        Gear  
...    ...        ...       ...         ...  
1899  1997         29       501        Gear  
1900  2007          0       501        Gear  
1901  2008          0       501        Gear  
1902  2021          0       501        Gear  
1903  2020          0       501        Gear  

[1904 rows x 6 columns]

**Q12:** Create a function called `theme_by_setnum` that takes as input the set_num (as a string) and returns the respective theme name as a string.

In [185]:
def theme_by_setnum(set_num):
    data_rename = themes.rename(columns={"id": "theme_id", "name":"name_themes"})
    data_merge = pd.merge(data_rename, sets, on='theme_id')
    setss = data_merge[data_merge['set_num'] == set_num]
    answer = setss['name_themes']
    return answer

**Code Check:** Visit the [Sets page](https://rebrickable.com/sets/) from the Rebrickable website and find a set that you like.  Call your `theme_by_setnum()` function on that set and make sure the correct theme name is output.

In [186]:
theme_by_setnum('002-1')

1    Technic
Name: name_themes, dtype: object

**Millennium Falcon Prep Exercise:** Find all the sets named 'Millennium Falcon'. Of these sets, find the set with the largest number of parts.  You will use this information for later questions.

In [154]:
millenniumfalcons = sets[sets['name'].str.contains('Millennium Falcon')].sort_values(by='num_parts',ascending=False).head(1)
millenniumfalcons

set_num               name  year  theme_id  num_parts
12884  75192-1  Millennium Falcon  2017       171       7541

In [155]:
inventory_minifigs

inventory_id     fig_num  quantity
0                 3  fig-001549         1
1                 4  fig-000764         1
2                19  fig-000555         1
3                25  fig-000574         1
4                26  fig-000842         1
...             ...         ...       ...
18139        122344  fig-009900         1
18140        122344  fig-009908         2
18141        122344  fig-009954         1
18142        122344  fig-009955         1
18143        122344  fig-009956         1

[18144 rows x 3 columns]

**Q13:** 
- How many minifigs are contained in the set from the `Millennium Falcon Prep Exercise`?  Save this as `Q13` (as an integer).  
- Note that there are sometimes multiple versions of a set as shown in the `inventories` table.  This will not matter for this question as there is only one version for the respective set but this should be noted for future questions.
- Remember to show your work or your answer will manually get marked as incorrect.

In [179]:
#get inventories dataframe ready to merge with the inventory_minifigs dataframe to then merge with the sets column by the 
# inventory_id and then the set_num column

rename_inventories = inventories.rename(columns={'id':'inventory_id'})
mergeddata1 =pd.merge(rename_inventories, inventory_minifigs, on='inventory_id')
mergeddata2 = pd.merge(mergeddata1, sets, on='set_num')
answer = mergeddata2[mergeddata2['set_num']== '75192-1']
Q13 = answer.shape[0]
Q13

8

**Q14:** 
- Create a function called `get_minifigs` that takes as input a set number (as a string) and a version number (as an integer) and outputs all minifigs from that set / version combination.  
- The DataFrame should be in `fig_num` order with the following columns/column names: `set_num`, `fig_num`, `fig_name`, `fig_num_parts`.  
- The index should go from 0 to n-1.

In [187]:
minifigs

fig_num                                   name  num_parts
0      fig-000001                     Toy Store Employee          4
1      fig-000002                           Customer Kid          4
2      fig-000003                  Assassin Droid, White          8
3      fig-000004                           Basic Figure          4
4      fig-000005        Captain America with Short Legs          3
...           ...                                    ...        ...
11822  fig-012221                                Bob-omb         11
11823  fig-012222                   Boomer (Banzai) Bill         10
11824  fig-012223                            Bullet Bill         10
11825  fig-012224                           Buzzy Beetle          7
11826  fig-012225  Unikitty - Cheerykitty (Cheery Kitty)         14

[11827 rows x 3 columns]

In [199]:
def get_minifigs(set_num, version):
    #rename columns to correctly identify them
    rename_inventories = inventories.rename(columns={'id':'inventory_id'})
    rename_minifigs = minifigs.rename(columns={'name':'fig_name','num_parts':'fig_num_parts'})
    #merge data with respective columns
    mergeddata1 =pd.merge(rename_inventories, inventory_minifigs, on='inventory_id')
    mergeddata2 = pd.merge(mergeddata1, rename_minifigs, on='fig_num')
    #extract the set_num and version
    minifigs1 = mergeddata2[(mergeddata2['set_num']== set_num) & (mergeddata2['version']== version)]
    minifigs1 = minifigs1[['set_num','fig_num','fig_name', 'fig_num_parts']].sort_values(by='fig_num')
    minifigs1 = minifigs1.reset_index(drop=True)
    return minifigs1

**Q15:** Using the above function, output information about the minifigs from the set discussed in the `Millennium Falcon Prep Exercise`.  Save this output as `Q15`.

In [200]:
Q15 = get_minifigs('75192-1', 1)
Q15

set_num     fig_num                                           fig_name  \
0  75192-1  fig-001714                              Chewbacca, Dark Brown   
1  75192-1  fig-001810                             Finn, Black Undershirt   
2  75192-1  fig-002057                                 Rey, Dark Tan Robe   
3  75192-1  fig-002514    C-3PO, Pearl Gold, Colorful Wires, Printed Legs   
4  75192-1  fig-002544                                               BB-8   
5  75192-1  fig-002549  Princess Leia, Hoth Outfit, White Jacket, Prin...   
6  75192-1  fig-002550  Han Solo, Dark Blue Jacket, Dark Brown Legs, P...   
7  75192-1  fig-002551                               Han Solo, Old, Angry   

   fig_num_parts  
0              3  
1              4  
2              4  
3              3  
4              2  
5              4  
6              4  
7              4

**Code Check:** Make sure that your `get_minifigs()` function works with both version 1 and version 2 of set 10217-1, Harry Potter's Diagon Alley.  Both versions should have 11 minifigs in the DataFrame.

In [197]:
get_minifigs('10217-1',1)

set_num     fig_num                                           fig_name  \
0   10217-1  fig-006084  Lucius Malfoy, Black Robes and Cape, Death Eat...   
1   10217-1  fig-002229                   Skeleton (Standard) (3626b Head)   
2   10217-1  fig-006075                 Rubeus Hagrid, Dark Brown Overcoat   
3   10217-1  fig-006130  Harry Potter, Open Dark Blue Jacket over White...   
4   10217-1  fig-006138                                    Fenrir Greyback   
5   10217-1  fig-006162  Ron Weasley, Red Tartan Sweater Vest, Reddish ...   
6   10217-1  fig-006163          Fred / George Weasley in Dark Orange Suit   
7   10217-1  fig-006164                                 Garrick Ollivander   
8   10217-1  fig-006165    Goblin, Black Jacket and Light Bluish Gray Legs   
9   10217-1  fig-006166        Goblin, Black Jacket and Reddish Brown Legs   
10  10217-1  fig-006167  Hermione Granger, Light Bluish Gray Sweater wi...   

    fig_num_parts  
0               5  
1               6  
2               3  
3               4  
4               4  
5               4  
6               4  
7               4  
8               3  
9               3  
10              4

In [198]:
get_minifigs('10217-1',2)

set_num     fig_num                                           fig_name  \
0   10217-1  fig-006084  Lucius Malfoy, Black Robes and Cape, Death Eat...   
1   10217-1  fig-002229                   Skeleton (Standard) (3626b Head)   
2   10217-1  fig-006075                 Rubeus Hagrid, Dark Brown Overcoat   
3   10217-1  fig-006130  Harry Potter, Open Dark Blue Jacket over White...   
4   10217-1  fig-006138                                    Fenrir Greyback   
5   10217-1  fig-006162  Ron Weasley, Red Tartan Sweater Vest, Reddish ...   
6   10217-1  fig-006163          Fred / George Weasley in Dark Orange Suit   
7   10217-1  fig-006164                                 Garrick Ollivander   
8   10217-1  fig-006165    Goblin, Black Jacket and Light Bluish Gray Legs   
9   10217-1  fig-006166        Goblin, Black Jacket and Reddish Brown Legs   
10  10217-1  fig-006167  Hermione Granger, Light Bluish Gray Sweater wi...   

    fig_num_parts  
0               5  
1               6  
2               3  
3               4  
4               4  
5               4  
6               4  
7               4  
8               3  
9               3  
10              4

**Q16:** 
- Create a function called `minifigs_from_themes` that takes as input a theme id (as an integer) and outputs all the minifigs contained in sets with that theme.  
- The DataFrame should be in `fig_num` order with the following columns/column names: `fig_num`, `fig_name`. 
- The index should go from 0 to n-1.  
- Exact duplicates should be removed so that only unique figs are included.

In [226]:
def minifigs_from_themes(theme_id):
    #rename appropiate columns in themes,sets, and inventories df
    themerenamed = themes.rename(columns={'id':'theme_id','name':'theme_name'})
    minifigsrenamed = minifigs.rename(columns ={'name':'fig_name','num_parts':'fig_num_parts'})
    inventoriesre = inventories.rename(columns={'id':'inventory_id'})
    setsrenamed = sets.rename(columns={'name':'sets_name','num_parts':'set_num_parts'})
    #merge all data together, with respect to columns in column
    data1 = pd.merge(themerenamed,setsrenamed, on='theme_id')
    data2 = pd.merge(inventoriesre, inventory_minifigs, on='inventory_id')
    data3 = pd.merge(data2,minifigsrenamed, on='fig_num')
    data4 = pd.merge(data3,data1, on='set_num')
    #call the theme_id 
    datareturn = data4[data4['theme_id']== theme_id]
    #call columns of importance
    datareturn = datareturn[['fig_num','fig_name']]
    #sort by fig_num
    datareturn = datareturn.sort_values(by='fig_num')
    #drop duplicates and reset the index
    datareturn = datareturn.drop_duplicates()
    datareturn = datareturn.reset_index(drop=True)
    #return datareturn
    return datareturn

**Code Check:** Call the `minifigs_from_themes()` function using the theme number of the set discussed in the `Millennium Falcon Prep Exercise`.  There should be 149 minifigs output for that respective theme.

In [227]:
minifigs_from_themes(171)

fig_num                                           fig_name
0    fig-000003                              Assassin Droid, White
1    fig-000124                           Chewbacca, Reddish Brown
2    fig-000144  Luke Skywalker, White Robe, White Legs, Cheek ...
3    fig-000303           Clone Trooper, Phase I Armor, Brown Eyes
4    fig-000516  Darth Vader, Light Bluish Gray Skin, Scarred R...
..          ...                                                ...
144  fig-011942  AT-AT Driver - Light Bluish Grey Helmet, Sand ...
145  fig-011944  Snowtrooper, White Hips, Reddish Brown Head, F...
146  fig-011945  Snowtrooper, White Hips, Light Flesh Head, Female
147  fig-011946  Snowtrooper, White Hips, Light Flesh Head, Off...
148  fig-011957    Snowtrooper, White Hips, Light Flesh Head, Grin

[149 rows x 2 columns]

**Q17:** 
- Create a function called `sets_from_minifig` that takes as input a minifig number (as a string) and returns a list of sets a minifig has appeared in.  
- Return a DataFrame (sorted by `set_num` and then by `version`) with the following columns: `set_num`, `version`,`name_sets`, `fig_num`, `fig_name`.  
- The index should go from 0 to n-1.

In [235]:
def sets_from_minifig(fig_num):
    #rename appropiate columns
    themerenamed = themes.rename(columns={'id':'theme_id','name':'theme_name'})
    minifigsrenamed = minifigs.rename(columns ={'name':'fig_name','num_parts':'fig_num_parts'})
    inventoriesre = inventories.rename(columns={'id':'inventory_id'})
    setsrenamed = sets.rename(columns={'name':'name_sets','num_parts':'set_num_parts'})
    #merge all data together, with respect to columns in column
    data1 = pd.merge(themerenamed,setsrenamed, on='theme_id')
    data2 = pd.merge(inventoriesre, inventory_minifigs, on='inventory_id')
    data3 = pd.merge(data2,minifigsrenamed, on='fig_num')
    data4 = pd.merge(data3,data1, on='set_num')
    data4 = data4[data4['fig_num']== fig_num]
    data4 = data4[['set_num','version','name_sets','fig_num','fig_name']].sort_values(by=['set_num','version']).reset_index(drop=True)
    return data4

**Code Check:** Call your `sets_from_minifig()` function on one of the Lego Friends minifigs, Andrea, who has figure number: `fig-001039`.  Your function should return a DataFrame that contains the following sets:

- Andrea on the Beach
- Birthday Party
- City Park Cafe

In [238]:
sets_from_minifig('fig-001039')

set_num  version            name_sets     fig_num  \
0  30100-1        1  Andrea on the Beach  fig-001039   
1  30107-1        1       Birthday Party  fig-001039   
2   3061-1        1       City Park Café  fig-001039   

                                            fig_name  
0  Andrea - Bright Light Orange Top, Light Aqua S...  
1  Andrea - Bright Light Orange Top, Light Aqua S...  
2  Andrea - Bright Light Orange Top, Light Aqua S...

**Q18:** 

*Note:  Before you begin working on this question, please note that CodeGrade has a 1GB RAM limit.  If you see an `Exit -9` code when submitting your assignment, it is probably due to this RAM limit and previous students had an issue with this question causing too much memory usage.  The biggest thing to think about is to only use the tables that you need to answer for this question.  For instance, you do not need any of the minifigs data since this is not needed for the output and the merging of this data could cause the number of rows to "explode" in terms of the total count.  I understand that this can be annoying that we have to work around CodeGrade's limitations, but it is good practice in memory management.*

*You should only need to merge the following datasets for this answer so as to not go above the RAM limit: `themes`, `sets`, `inventories`, `inventory_parts`, `parts`, and `colors`*

- Create a function called `set_parts` that takes as input the set number (as a string) and version number (as an integer) and outputs a DataFrame with each part in the set.  
- The DataFrame should have the following columns/column names: `set_num`, `version`, `name_sets`, `year`, `name_themes`, `part_num`, `name_part`, `name_color`,`quantity`,`is_spare`.  
- The Data should be sorted by `part_num` and then `name_color`.  
- The index should go from 0 to n-1.

In [166]:
def set_parts(set_num,version):
    #rename appropiate columns
    renasets = sets.rename(columns={'name':'name_sets'})
    renathemes = themes.rename(columns={'name':'name_themes','id':'theme_id'})
    renaparts = parts.rename(columns={'name':'name_part'})
    renacolors = colors.rename(columns={'id':'color_id','name':'name_color'})
    renainventories = inventories.rename(columns={'id':'inventory_id'})
    #merge dataframes by columns
    data1 = pd.merge(renasets, renathemes, on='theme_id')
    data2 = pd.merge(data1, renainventories, on='set_num')
    data3 = pd.merge(data2,inventory_parts, on='inventory_id')
    data4 = pd.merge(data3, renaparts, on='part_num')
    data5 = pd.merge(data4, renacolors, on='color_id')
    data5 = data5[(data5['set_num']== set_num) & (data5['version']== version)]
    data6 = data5[['set_num','version','name_sets','year','name_themes','part_num','name_part','name_color','quantity','is_spare']]
    data6 = data6.sort_values(by=['part_num','name_color'])
    data6 = data6.reset_index(drop=True)
    return data6

**Code Check**: Using the `set_parts()` function, output a list of parts for the set discussed in the `Millennium Falcon Prep Exercise`. There should be a total of 730 rows in the DataFrame, which includes 676 parts and 54 spare parts. You can check the respective page on the rebrickable website to verify these results.

In [167]:
set_parts('75192-1',1)

set_num  version          name_sets  year                name_themes  \
0    75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
1    75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
2    75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
3    75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
4    75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
..       ...      ...                ...   ...                        ...   
725  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
726  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
727  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
728  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
729  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   

    part_num                                          name_part  \
0      10197  Technic Pin Connector Hub with 2 Perpendicular...   
1      10201            Bracket 1 x 2 - 1 x 4 [Rounded Corners]   
2      10247  Plate Special 2 x 2 with 1 Pin Hole [Complete ...   
3      10247  Plate Special 2 x 2 with 1 Pin Hole [Complete ...   
4      10314  Brick Curved 1 x 4 x 1 1/3 No Studs, Curved To...   
..       ...                                                ...   
725    99563  Tile Special 1 x 2 with Sloped Walls AKA Money...   
726    99780                     Bracket 1 x 2 - 1 x 2 Inverted   
727    99780                     Bracket 1 x 2 - 1 x 2 Inverted   
728    99781                              Bracket 1 x 2 - 1 x 2   
729    99784  Bar 1 x 12 with 1 x 2 Plate End with Hollow St...   

            name_color  quantity is_spare  
0     Dark Bluish Gray         3        f  
1    Light Bluish Gray        12        f  
2     Dark Bluish Gray         5        f  
3                White        22        f  
4                  Tan         3        f  
..                 ...       ...      ...  
725  Light Bluish Gray        70        f  
726              Black         6        f  
727   Dark Bluish Gray        25        f  
728  Light Bluish Gray         2        f  
729  Light Bluish Gray         3        f  

[730 rows x 10 columns]

## Optional Exercises
These exercises are optional and are not graded by CodeGrade.  They can be used as extra optional practice.

- What is the median year for the sets data?  Given the oldest and newest year in the data, what does this median value tell you about whether the number of sets have increased or decreased over the years?
- What minifig has the most number of parts?
- What minifig shows up in the most number of sets?
- When was the first minifig included in a set?
- What set has the most number of minifigs
- What part is used the most for all sets? 
- What color is used the most for all parts?
- Get a list of all unique color names, sort by color name, and output as a list. 
- How many total transparent colors are in the colors DataFrame?
- Create a function called get_part_colors that takes as input a part number (as a string) and returns a DataFrame of all colors a part has appeared in.
- What part / color combination is used the most for all sets?
- What part / color combination is used the least for all sets?
- Explore some of your favorite themes and which colors are used the most for those themes.
- Plot a line graph of the average number of parts per set per year.
